### PR TITLE
[SYCL] Add missing relational functions + bunch of other fixes

### PR DESCRIPTION
**Pull request by dm-vodopyanov**
This patch makes math_builtin_api SYCL CTS test compilable and fixes most of the run failures.

1. Add missing marray relational functions: `any`, `all`, `bitselect`
2. Fix scalar `select` relational function: it didn't work for most of the types
3. Align scalar `abs` integer function with SYCL 2020: now it returns signed integer type instead of unsigned for signed argument
4. Fix math functions which take multi_ptr argument: now they accept multi_ptr without decoration in accordance with SYCL 2020

### CODE REVIEW

File: sycl/test-e2e/Basic/built-ins/marray_math.cpp
Comment (dm-vodopyanov) [inline sycl/test-e2e/Basic/built-ins/marray_math.cpp:47]: Note for reviewers: changed here to `no` just to check explicitly the case without decoration, and case with decoration will be checked implicitly here

File: sycl/include/sycl/builtins.hpp
Comment (dm-vodopyanov) [inline sycl/include/sycl/builtins.hpp:526]: Tag @steffenlarsen in case you have better solution

Comment (dm-vodopyanov) [reply sycl/include/sycl/builtins.hpp:526]: Steffen proposed to use `get_decorated()` here 

https://github.com/intel/llvm/blob/09ceb1034f04ef7156c7ea2be0f3cd1bcbc12df2/sycl/include/sycl/detail/generic_type_traits.hpp#L391 

for pointers `t` with `decorated::no`. Will try tomorrow. It will convert non-decorated pointer to decorated  - the same logic as it is now in this PR, but the implementation should look much cleaner.

Comment (aelovikov-intel) [reply sycl/include/sycl/builtins.hpp:526]: But it would be plain pointer (`__global int *`), not `multi_ptr`. What am I missing?

File: sycl/include/sycl/builtins.hpp
Comment (aelovikov-intel) [inline sycl/include/sycl/builtins.hpp:525]: How would it work if `iptr` is something like `__global float *`? I'd expect we need to 
```
if constexpr(is_multi_ptr) 
  if constexpr (iptr.is_decorated)
```

File: sycl/include/sycl/builtins.hpp
Comment (aelovikov-intel) [inline sycl/include/sycl/builtins.hpp:1129]: Why do we change return type here? If it's valid change, then the comment on line 1127 needs to be updated.

File: sycl/include/sycl/builtins.hpp
Comment (aelovikov-intel) [inline sycl/include/sycl/builtins.hpp:2130]: Should we do `sizeof(...)`-based checks instead?

File: sycl/include/sycl/builtins.hpp
Comment (aelovikov-intel) [inline sycl/include/sycl/builtins.hpp:2223]: What is this line for?

File: sycl/include/sycl/builtins.hpp
Comment (aelovikov-intel) [inline sycl/include/sycl/builtins.hpp:2217]: `marray` has `begin`/`end`, so we can use `std::any` (preferably) or at least range-based for loop.

File: sycl/include/sycl/builtins.hpp
Comment (AlexeySachkov) [reply sycl/include/sycl/builtins.hpp:526]: @aelovikov-intel, the built-in accepts `genfloatptr` in accordance with the SYCL spec, which is in turn defined as `multir_ptr` in [Table 165](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#table.gentypes)

> genfloatptr
> All permutations of multi_ptr<DataT, AddressSpace, IsDecorated> where DataT is all types within genfloat, AddressSpace is access::address_space::global_space, access::address_space::local_space and access::address_space::private_space and IsDecorated is access::decorated::yes and access::decorated::no.

Essentially, you can't pass a raw pointer to the built-in. But I would double-check if our definition of `is_genfloatptr` matches the spec

Comment (dm-vodopyanov) [reply sycl/include/sycl/builtins.hpp:526]: > `is_genfloatptr`

Correct, it doesn't match the spec. I keep in mind that the separate patch should be made which deprecates current scalar functions which accept decorated::legacy, but this is second priority, so didn't introduce it in this PR.

File: sycl/include/sycl/builtins.hpp
Comment (dm-vodopyanov) [reply sycl/include/sycl/builtins.hpp]: Applied in https://github.com/intel/llvm/pull/9110/commits/dd017157c35bf89331f71cf1799b1e083a46330d

File: sycl/include/sycl/builtins.hpp
Comment (dm-vodopyanov) [reply sycl/include/sycl/builtins.hpp]: Thanks, doesn't need anymore, removed

File: sycl/include/sycl/builtins.hpp
Comment (dm-vodopyanov) [reply sycl/include/sycl/builtins.hpp:1087]: This is according to SYCL 2020. Updated line 1127

**Comment by dm-vodopyanov** [comment]
**SYCL Pre Commit / Linux / L0 GEN9 LLVM Test Suite (pull_request_target)** failure is unrelated - CI issue, submitted https://github.com/intel/llvm/issues/9140. Will try to re-start this job.

File: sycl/include/sycl/detail/generic_type_traits.hpp
Comment (aelovikov-intel) [inline sycl/include/sycl/detail/generic_type_traits.hpp:410]: I think this should go to somewhere near https://github.com/intel/llvm/blob/114906b688f912f58a2e90901b92ce18db37b999/sycl/include/sycl/detail/type_traits.hpp#L283-L293 with corresponding change in naming convention. Also, I think it would be better to introduce a simpler `IsMultiPtr` together with these more fine-grained traits.

I'd also like to see `is_*multi_ptr_v` helpers to be implemented here.

File: sycl/include/sycl/builtins.hpp
Comment (dm-vodopyanov) [reply sycl/include/sycl/builtins.hpp:2094]: I thought this is not reliable based on Alexey's comment, so I closed it. Should we do sizeof-based checks in such case?

File: sycl/include/sycl/builtins.hpp
Comment (dm-vodopyanov) [reply sycl/include/sycl/builtins.hpp:1090]: I tried and faced with some incompatibility with HIP and libclc, so instead changed the SYCL high-level API. Changing SYCL API (and not OpenCL's __invoke***) also looks cleaner for me, because there is no OpenCL's abs builtin which returns signed type.

File: sycl/include/sycl/builtins.hpp
Comment (dm-vodopyanov) [reply sycl/include/sycl/builtins.hpp:2176]: Yes, bool should be here as this is this is pure SYCL 2020 interface, and detail::anyall_ret_t was created for supporting both 1.2.1 and 2020 versions.

File: sycl/include/sycl/builtins.hpp
Comment (dm-vodopyanov) [reply sycl/include/sycl/builtins.hpp:1091]: This PR fixes lots of CTS tests and we need to collect internal test results as soon as possible - so this change can be made in a separate PR

File: sycl/include/sycl/builtins.hpp
Comment (dm-vodopyanov) [reply sycl/include/sycl/builtins.hpp]: Misunderstood this, sorry, implemented.

File: sycl/include/sycl/builtins.hpp
Comment (dm-vodopyanov) [reply sycl/include/sycl/builtins.hpp]: Applied.

File: sycl/include/sycl/detail/generic_type_traits.hpp
Comment (dm-vodopyanov) [reply sycl/include/sycl/detail/generic_type_traits.hpp]: Applied.

File: sycl/test-e2e/Basic/built-ins/marray_relational.cpp
Comment (aelovikov-intel) [inline sycl/test-e2e/Basic/built-ins/marray_relational.cpp:43]: I'd use

```
buffer<TYPE> b{1};
q.submit();
assert(host_accessor{b}[0] == EXPECTED[0]);
```

to reduce indentation.

File: sycl/include/sycl/builtins.hpp
Comment (aelovikov-intel) [inline sycl/include/sycl/builtins.hpp:2080]: I'd prefer `get_select_opencl_builtin_c_arg_type` to be inlined to here.

**Review by aelovikov-intel — APPROVED** [review]
LGTM assuming the "c_type" helper will be inlined before merging this and @AlexeySachkov doesn't have any concerns on his side.

File: sycl/include/sycl/builtins.hpp
Comment (dm-vodopyanov) [reply sycl/include/sycl/builtins.hpp]: Applied.

File: sycl/test-e2e/Basic/built-ins/marray_relational.cpp
Comment (dm-vodopyanov) [reply sycl/test-e2e/Basic/built-ins/marray_relational.cpp:43]: All these marray***.cpp tests will be refactored to get rid of macros in the separate patch, this comment will be applied.

**Comment by dm-vodopyanov** [comment]
Failure in USM/usm_pooling.cpp is unrelated: https://github.com/intel/llvm/issues/9148

**Pull request merged**
